golem-worker-executor-base/src/worker.rs

```diff
@@ -86,6 +89,41 @@ pub struct Worker<Ctx: WorkerCtx> {
     oom_retry_config: RetryConfig,
 }
 
+impl<Ctx: WorkerCtx> Worker<Ctx> {
+    pub async fn get_component_metadata(
+        worker: &Arc<Worker<Ctx>>,
+    ) -> Result<(WorkerMetadata, Component, ComponentMetadata), GolemError> {
+        let component_id = worker.owned_worker_id.component_id();
+        let worker_metadata = worker.get_metadata().await?;
+
+        let component_version = worker_metadata
+            .last_known_status
+            .pending_updates
+            .front()
+            .map_or(
+                worker_metadata.last_known_status.component_version,
+                |update| {
+                    let target_version = *update.description.target_version();
+                    info!(
+                        "Attempting {} update from {} to version {target_version}",
+                        match update.description {
+                            UpdateDescription::Automatic { .. } => "automatic",
+                            UpdateDescription::SnapshotBased { .. } => "snapshot based",
+                        },
+                        worker_metadata.last_known_status.component_version
+                    );
+                    target_version
+                },
+            );
+        let (component, component_metadata) = worker
+            .component_service()
+            .get(&worker.engine(), &component_id, component_version)
```

Comment:
This is potentially download, compiling and returning the whole WASM component, not just the metadata! Must not be part of this function, only done once in the beginning of the invocation loop

Reply:
Why is it a download @vigoo ?
In the invoke functions in grpc, I am calling this after `start_if_needed`. So we should be using the cache by the time we call this function I suppose?

Reply:
It's not an infinite cache. The currently configured component cache size is 32, while the component metadata cache size is 16k. And there is absolutely no need to return the `Component` (which is the compiled wasmtime thing!) in a function that is called `get_component_metadata`.

Reply:
Ok I understand. I fixed this bit. 

Yes, I understand it cannot be infinite cache.  I have changed this to use only component metadata cache and hopefully that goes to 16K elements cache where we could possibly have the metadata already downloaded once the instance is started (if it decides to start) and the next subsequent call of get_metadata will hopefully hit the cache. That's my assumption. Ofcourse if there is a `possible` download , but just wanted to ensure the miss rate is possibly high.

The new change:
```scala
 async fn invoke_worker_internal<Req: GrpcInvokeRequest>(
        &self,
        request: &Req,
    ) -> Result<(), GolemError> {
        let full_function_name = request.name();

        let calling_convention = request.calling_convention();

        let worker = self.get_or_create(request).await?;

        let component_metadata = worker::get_component_metadata(&worker).await?;
           ......
           
            .....
           
    }

```

Internally I can see it hits only component_metadata_cache, so hopefully this change would have fixed your concern.

```scala

 async fn get_metadata(
        &self,
        component_id: &ComponentId,
        forced_version: Option<ComponentVersion>,
    ) -> Result<ComponentMetadata, GolemError> {
        match forced_version {
            Some(version) => {
                let client = self.client.clone();
                let access_token = self.access_token;
                let retry_config = self.retry_config.clone();
                let component_id = component_id.clone();
                self.component_metadata_cache
                    .get_or_insert_simple(
                        &ComponentKey {
                            component_id: component_id.clone(),
                            component_version: version,
                        },
                        || {
                            Box
```